base/sparse/sparsevector.jl

```diff
@@ -0,0 +1,1441 @@
+"""
+function sparsevec{Tv,Ti<:Integer}(dict::Associative{Ti,Tv})
+    m = length(dict)
+    nzind = Array(Ti, m)
+    nzval = Array(Tv, m)
+
+    cnt = 0
+    len = zero(Ti)
+    for (k, v) in dict
+        k >= 1 || throw(ArgumentError("index must be positive."))
+        if k > len
+            len = k
+        end
+        if v != 0
+            cnt += 1
+            @inbounds nzind[cnt] = k
+            @inbounds nzval[cnt] = v
+        end
+    end
+    resize!(nzind, cnt)
+    resize!(nzval, cnt)
+    _sparsevector!(nzind, nzval, len)
+end
+
+function sparsevec{Tv,Ti<:Integer}(dict::Associative{Ti,Tv}, len::Integer)
+    m = length(dict)
+    nzind = Array(Ti, m)
+    nzval = Array(Tv, m)
+
+    cnt = 0
+    maxk = convert(Ti, len)
+    for (k, v) in dict
+        1 <= k <= maxk || throw(ArgumentError("an index (key) is out of bound."))
+        if v != 0
+            cnt += 1
+            @inbounds nzind[cnt] = k
+            @inbounds nzval[cnt] = v
+        end
+    end
+    resize!(nzind, cnt)
+    resize!(nzval, cnt)
+    _sparsevector!(nzind, nzval, len)
+end
+
+
+### Element access
+
+function setindex!{Tv,Ti<:Integer}(x::SparseVector{Tv,Ti}, v::Tv, i::Ti)
+    checkbounds(x, i)
+    nzind = nonzeroinds(x)
+    nzval = nonzeros(x)
+
+    m = length(nzind)
+    k = searchsortedfirst(nzind, i)
+    if 1 <= k <= m && nzind[k] == i  # i found
+        if v == 0
+            deleteat!(nzind, k)
+            deleteat!(nzval, k)
+        else
+            nzval[k] = v
+        end
+    else  # i not found
+        if v != 0
+            insert!(nzind, k, i)
+            insert!(nzval, k, v)
+        end
+    end
+    x
+end
+
+setindex!{Tv, Ti<:Integer}(x::SparseVector{Tv,Ti}, v, i::Integer) =
+    setindex!(x, convert(Tv, v), convert(Ti, i))
+
+
+### Conversion
+
+# convert SparseMatrixCSC to SparseVector
+function convert{Tv,Ti<:Integer}(::Type{SparseVector{Tv,Ti}}, s::SparseMatrixCSC{Tv,Ti})
+    size(s, 2) == 1 || throw(ArgumentError("The input argument must have a single-column."))
+    SparseVector(s.m, s.rowval, s.nzval)
+end
+
+convert{Tv,Ti}(::Type{SparseVector{Tv}}, s::SparseMatrixCSC{Tv,Ti}) =
+    convert(SparseVector{Tv,Ti}, s)
+
+convert{Tv,Ti}(::Type{SparseVector}, s::SparseMatrixCSC{Tv,Ti}) =
+    convert(SparseVector{Tv,Ti}, s)
+
+# convert Vector to SparseVector
+
+"""
+    sparsevec(A)
+
+Convert a vector `A` into a sparse vector of length `m`.
+"""
+sparsevec{T}(a::AbstractVector{T}) = convert(SparseVector{T, Int}, a)
+sparsevec(a::AbstractArray) = sparsevec(vec(a))
+sparsevec(a::AbstractSparseArray) = vec(a)
+sparse(a::AbstractVector) = sparsevec(a)
+
+function _dense2sparsevec{Tv,Ti}(s::AbstractArray{Tv}, initcap::Ti)
+    # pre-condition: initcap > 0; the initcap determines the index type
+    n = length(s)
+    cap = initcap
+    nzind = Array(Ti, cap)
+    nzval = Array(Tv, cap)
+    c = 0
+    @inbounds for i = 1:n
+        v = s[i]
+        if v != 0
+            if c >= cap
+                cap *= 2
```

Comment:
Did you try using `sizehint!(nzind, cap)`, etc?  I agree it'd be better to use push here, but thanks for benchmarking it!